### PR TITLE
Remove projections from app activation

### PR DESCRIPTION
**Pull request by bviglietta**
There are a couple of places where we expose projection classes in our public interface. One of them is in the app activation path: if the app is activated via some protocol, then a projection class will be delivered to the app delegate. This change removes those uses of projection classes. This is a **breaking change** for clients.

There are several different kinds of activation that we support.

1. Toast notification: This never used projection classes, so there's no change here.

2. Protocol (i.e., URL) activation: This previously delivered a WFUri* to the app and will now deliver a NSURL*.

3. Voice command activation
4. File activation
These two delivered a WMSSpeechRecognitionResult* and a WAAFileActivatedEventArgs*, respectively. Now they will both deliver a RTObject*, the base class of all projection classes. In order to get to the derived type, client code will have to rt_dynamic_cast the object to the appropriate type.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
Due to the nature of Objective-C, it's very easy for us to do this. We can just pass along an `NSObject<SomeProtocolHere>*` and hide a proxy class inside UIKit.

**Comment by jaredhms** [comment]
@DHowett-MSFT - this change lets them either use projections OR CppWinRT, but I agree it's a bit messy.  I don't understand the 'internal crap' comment though - these are actual WinRT types that the app needs to know about and work with - definitely not internal to us.  

My main concern is that RTObject* loses all type info - maybe we just add placeholder classes that are documented to be 'cast'-able to either their Projections OR CppWinRT counterparts.

**Comment by bviglietta** [comment]
Dustin, I'm not sure I understand all of your points here. We already do implement these activations as application launch with options: they show up as entries inside the options dictionary that the app delegate receives in application:didFinishLaunchingWithOptions:.

Also, why is it burdensome to use the projection classes? I don't see how WMSSpeechRecognitionResult is more difficult to use than NSObject<SpeechRecognitionResultProtocolThatIJustInvented>.

I'll grant that the rt_dynamic_cast requirement is a new thing that this change adds, but I think it's pretty minor.

**Comment by bviglietta** [comment]
If people really don't like the typeless RTObjects, I'm entirely open to introducing some new classes -- SpeechActivation and FileActivation, let's say -- that wrap the underlying WinRT objects.

**Comment by DHowett-MSFT** [comment]
> this change lets them either use projections OR CppWinRT

Not exactly; we're not exposing CppWinRT across the boundary, so they'd have to unbox RTObject and that is not supported externally. We have `comObj` internally for unboxing.

> "internal crap"

Yes and no. They're projected objects, and their contents are important, but we are forcing them to take a projections dependency to use them. It leaks an implementation detail across the API boundary. If we go for a protocol (specifically: that gives us full agency as to internal implementation, where a publically exposed class name does not), we can change things as we wish without betraying internal details.

> We already do implement these activations as application launch with options

We're not, though. We augmented the UIApplicationDelegate protocol with these three callbacks.

```objc
- (void)application:(UIApplication*)application didReceiveVoiceCommand:(WMSSpeechRecognitionResult*)result;
- (void)application:(UIApplication*)application didReceiveFile:(WAAFileActivatedEventArgs*)result;
- (void)application:(UIApplication*)application didReceiveProtocol:(WFUri*)uri;
```

I don't think it's bad that we did that -- after all, we've greatly expanded its capabilities -- but it doesn't fit the application activation model from the reference platform.

As an iOS developer, I hated having to bifurcate my app initialization across the iOS 4 and iOS 5 and iOS 6 initializers. It's no different when _we_ add new ones.

> more burdensome

You're right. I think it's more mentally taxing than truly taxing, as on both sides of the implementation we/they have to care about precisely what variety of object it is.

> guaranteed to be "cast"-able

protocols, mang! it'll rock our world.

**Comment by bviglietta** [comment]
> We're not, though [implementing these as application launch options].

We are, really! The methods you pointed out that we added to the application delegate are useful if an activation comes in after the app is already running.

File: Frameworks/UIKit/UIApplicationMain.mm [hunk 1]
```diff
@@ -336,7 +340,7 @@ void _UIApplicationShutdown() {
     THROW_NS_IF_FAILED(comPtr.As(&map));
 
     NSMutableDictionary* userInput = nil;
-    THROW_NS_IF_FAILED(Collections::WRLToNSCollection(map, &userInput));
+    THROW_NS_IF_FAILED(::Collections::WRLToNSCollection(map, &userInput));
```
Comment (aballway) [inline Frameworks/UIKit/UIApplicationMain.mm]: Is there a need for this to be changed?

**Comment by DHowett-MSFT** [comment]
> The methods you pointed out that we added to the application delegate are useful if an activation comes in after the app is already running.

Cool! I didn't consider that use case.

I don't want to be too bullish on the compatibility front. I know that we're still a pre-release and at 0.2... but I think it's important for us to come up with a way forward that maintains our agency while exposing an Objective-C or Objective-C++ facade that doesn't require the consumer to worry too deeply about _how_ we've done something.

If we're going to make a breaking change, I guess, let's do it once and let's do it right? 😸

File: tests/functionaltests/Tests/CortanaTests.mm [hunk 2]
```diff
@@ -277,6 +277,13 @@ - (void)application:(UIApplication*)application didReceiveProtocol:(WFUri*)uri {
                 return S_OK;
             });
```
Comment (aballway) [inline tests/functionaltests/Tests/CortanaTests.mm]: nit: can ToString be safely removed?

File: Frameworks/UIKit/UIApplicationMain.mm (hunk 1 above)
Comment (bviglietta) [inline Frameworks/UIKit/UIApplicationMain.mm]: Yes, it's to disambiguate between global Collections and Windows::Foundation::Collections.

---
In reply to: [112285489](https://github.com/Microsoft/WinObjC/pull/2569#discussion_r112285489) [](ancestors = 112285489)

File: tests/functionaltests/Tests/CortanaTests.mm (hunk 2 above)
Comment (bviglietta) [inline tests/functionaltests/Tests/CortanaTests.mm]: No, ToString is also used.

---
In reply to: [112285907](https://github.com/Microsoft/WinObjC/pull/2569#discussion_r112285907) [](ancestors = 112285907)

**Comment by bviglietta** [comment]
In the latest commit, I've switched back to using the projection types, so this is no longer a breaking change. The only difference between this and the original code is that we now load the projection classes dynamically, so there's no link-time dependency between UIKit and the projections.

(I even switched protocol activation back to WFUri* from NSURL*, even though NSURL was arguably a strict improvement.)

File: Frameworks/UIKit/UIApplicationMain.mm [hunk 3]
```diff
@@ -352,12 +355,19 @@ void _UIApplicationShutdown() {
 }
 
 extern "C" void UIApplicationMainHandleVoiceCommandEvent(IInspectable* voiceCommandResult) {
-    WMSSpeechRecognitionResult* result = [WMSSpeechRecognitionResult createWith:voiceCommandResult];
-    [[UIApplication sharedApplication] _sendVoiceCommandReceivedEvent:result];
+    id result = _createProjectionObject("WMSSpeechRecognitionResult", voiceCommandResult, "voice commands");
+
+    if (result) {
```
Comment (jaredhms) [inline Frameworks/UIKit/UIApplicationMain.mm]: perhaps we should fire the event even if 'result' is nil, so it's more clear to the app dev that they need to pull in projections?

File: Frameworks/UIKit/UIApplicationMain.mm [hunk 4]
```diff
@@ -384,8 +394,27 @@ void _UIApplicationShutdown() {
...
+    id result = _createProjectionObject("WFUri", protocolUri, "protocol activation");
+
+    if (result) {
+        NSString* source = _bundleIdFromPackageFamilyName(sourceApplication);
+
+        [[UIApplication sharedApplication] _sendProtocolReceivedEvent:result source:source];
+    }
+}
+
+id _createProjectionObject(const char* className, IInspectable* source, const char* description) {
```
Comment (jaredhms) [inline Frameworks/UIKit/UIApplicationMain.mm]: >_createProjectionObject [](start = 3, length = 23)

should we switch to this approach for UIView's public xamlElement property, and for initWith:xamlElement?

Comment (bviglietta) [inline Frameworks/UIKit/UIApplicationMain.mm]: Seems like a good idea.

---
In reply to: [112307470](https://github.com/Microsoft/WinObjC/pull/2569#discussion_r112307470) [](ancestors = 112307470)

File: Frameworks/UIKit/UIApplicationMain.mm (hunk 3 above)
Comment (bviglietta) [inline Frameworks/UIKit/UIApplicationMain.mm]: Maybe so. If we do that, then we should also, for consistency, put an NSNull into the launchOptions dictionary.

---
In reply to: [112307083](https://github.com/Microsoft/WinObjC/pull/2569#discussion_r112307083) [](ancestors = 112307083)

File: Frameworks/UIKit/UIApplicationMain.mm [hunk 5]
```diff
@@ -384,8 +394,27 @@ void _UIApplicationShutdown() {
...
+    // We can't refer to projection classes directly, because we don't want UIKit to have any link-time
+    // dependencies on the projections. However, we can try to load the classes dynamically.
+
+    id result = nil;
+    id cls = objc_getClass(className);
+
+    if (cls) {
+        result = objc_msgSend(cls, @selector(createWith:), source);
```
Comment (DHowett-MSFT) [inline Frameworks/UIKit/UIApplicationMain.mm]: nit: `id objc_msgSend(id, SEL, ...)` does not guarantee that vararg calling convention matches that of fixed arguments. This requires a cast to be completely safe.

Alternatively,
```objc
@protocol _ProjectionClassInformal
+ (id)createWith:(IInspectable*)i;
@end
// ...
result = [static_cast<Class<_ProjectionClassInformal>>(cls) createWith:source]
```

**Comment by jaredhms** [comment]
:shipit:

**Comment by bbowman** [comment]
:shipit:

**Pull request merged**